File: traveling_salesman_problem/rust/adjacency_matrix.rs
use std::iter::{repeat};
use std::ops::{Index, IndexMut};

use graph::{Edge, Graph};

pub struct AdjacencyMatrix {
    rows: Vec<Vec<u32>>,
}

impl AdjacencyMatrix {
    fn new(size: usize) -> Self {
        let row = vec![0; size];
        let rows = repeat(row).take(size).collect();

        AdjacencyMatrix { rows: rows }
    }

    pub fn from_graph(g: &Graph) -> Self {
        let mut matrix = AdjacencyMatrix::new(g.vertex_count);

        for &Edge { x, y, distance } in &g.edges {
            matrix[x][y] = distance;
            matrix[y][x] = distance;
        }

        matrix
    }

    pub fn size(&self) -> usize {
        self.rows.len()
    }

    pub fn distance(&self, tour: &Vec<usize>) -> u32 {
        // TODO: change to reduce
        let mut distance = 0;
        for i in 0..tour.len() {
            let j = if i < tour.len() - 1 { i + 1 } else { 0 };
            distance += self[tour[i]][tour[j]];
        }
        distance
    }
}

// For convenience, enable writing matrix[x] instead of matrix.rows[x].

impl Index<usize> for AdjacencyMatrix {
    type Output = Vec<u32>;

    fn index<'a>(&'a self, _index: usize) -> &'a Vec<u32> {
        &self.rows[_index]
    }
}

impl IndexMut<usize> for AdjacencyMatrix {
    fn index_mut<'a>(&'a mut self, _index: usize) -> &'a mut Vec<u32> {
        &mut self.rows[_index]
    }
}

File: traveling_salesman_problem/typescript/adjacency_matrix.ts
export interface Graph {
    vertexCount: number;
    edges: Edge[];
}

export interface Edge {
    x: number;
    y: number;
    distance: number;
}

export class AdjacencyMatrix {
    m: number[][];

    constructor(graph: Graph) {
        this.m = constructAdjacencyMatrix(graph);
    }

    tourDistance(tour: number[]) {
        return tourDistance(this.m, tour);
    }
}

function constructAdjacencyMatrix(graph: Graph): number[][] {
    const m = Array.from(
        {length: graph.vertexCount},
        () => new Array(graph.vertexCount)
    );

    for (const edge of graph.edges) {
        insertEdge(m, edge);
    }

    return m;
}

function insertEdge(m: number[][], edge: Edge): void {
    const {x, y, distance} = edge;
    m[x][y] = distance;
    m[y][x] = distance;
}

function tourDistance(m: number[][], tour: number[]): number {
    return tour.length < 2 ? 0 : tour.reduce((d, x, i) => {
        const j = (i + 1) % tour.length;
        const y = tour[j];
        return d + m[x][y];
    }, 0);
}
